src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> tempNode = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(tempNode.prev, value, tempNode);
+        if (index == 0) {
+            head = newNode;
+        } else {
+            tempNode.prev.next = newNode;
+        }
+        tempNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validIndex(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validIndex(index);
+        T t = findNodeByIndex(index).item;
+        findNodeByIndex(index).item = value;
+        return t;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validIndex(index);
+        T t = findNodeByIndex(index).item;
+        unlink(findNodeByIndex(index));
+        return t;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> node = head; node != null; node = node.next) {
+            if (node.item == object || object != null && object.equals(node.item)) {
+                unlink(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void validIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out off list length " + size);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out off list length " + size);
+        }
```

Comment:
modify yout validIndex() method and use it instead of hardcoding
for example change method signature to "private Node<T> findNodeByIndex(int index, int limit)"

Reply:
Such a check is used only in this method and it is more visual. Then findNodeByIndex does not need to pass two parameters.